benchmarks/helics/BenchmarkFederate.hpp

```diff
@@ -0,0 +1,334 @@
+#include "helics/helics-config.h"
+#include "helics_benchmark_util.h"
+
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+using namespace helics;
+/** class implementing common functionality for benchmarks */
+class BenchmarkFederate {
+  public:
+    enum class OutputFormat {
+        PLAIN_TEXT,
+        //          JSON
+    };
+
+  public:
+    // getters and setters for parameters
+    /** sets the delta time parameter
+     * @param dt the delta time to set
+     */
+    void setDeltaTime(helics::Time dt) { deltaTime = dt; }
+    /** gets the delta time parameter*/
+    helics::Time getDeltaTime() { return deltaTime; }
+    /** sets the final time parameter
+     * @param t the final/stop time*/
+    void setFinalTime(helics::Time t) { finalTime = t; }
+    /** gets the final time parameter*/
+    helics::Time getFinalTime() { return finalTime; }
+
+    /** sets the index parameter
+     * @param i the index
+     */
+    void setIndex(int i) { index = i; }
+    /** gets the index parameter*/
+    int getIndex() { return index; }
+    /** sets the max index parameter
+     * @param i the max index
+     */
+    void setMaxIndex(int i) { maxIndex = i; }
+    /** gets the max index parameter*/
+    int getMaxIndex() { return maxIndex; }
+
+    // functions for setting callbacks
+    /** sets a callback function to call immediately after doMainLoop() returns, but before
+     * the helics finalize() call
+     * @param cb a function that takes no arguments and returns void
+     */
+    void setBeforeFinalizeCallback(std::function<void()> cb = {}) { callBeforeFinalize = cb; }
+    /** sets a callback function to all after the helics finalize() call completes
+     * @param cb a function that takes no arguments and returns void
+     */
+    void setAfterFinalizeCallback(std::function<void()> cb = {}) { callAfterFinalize = cb; }
+
+    /** sets the output format to use when printing results
+     * @param f format to print results in
+     */
+    void setOutputFormat(OutputFormat f) { result_format = f; }
+
+    // protected to give derived classes more control
+  protected:
+    class Result {
+      public:
+        std::string name;
+        std::string key;
+        std::string value;
+    };
+    std::vector<Result> results; //!< a vector of output results to print
+
+    OutputFormat result_format = OutputFormat::PLAIN_TEXT; //!< output format for printing results
+
+    // parameters most benchmark federates need
+    helics::Time deltaTime = helics::Time(10, time_units::ns); //<! sampling rate
+    helics::Time finalTime = helics::Time(10000, time_units::ns); //<! final time
+    int index = 0; //<! the index for an instance of the benchmark federate
+    int maxIndex = 0; //<! the maximum index + 1 given to a benchmark federate in a run
+
+    // CLI11 Options for derived classes to change them if needed (e.g. set required)
+    CLI::Option* opt_delta_time; //<! the CLI11 option for --delta_time
```

Comment:
these should be initialized to nullptr since I believe we have a default constructor

Reply:
The no arg constructor calls the one that creates the CLI11 app using an empty-string for the name.